bioimageio/core/prediction_pipeline/_combined_processing.py

```diff
@@ -0,0 +1,183 @@
+        ]
+
+        # There is a difference between pre-and-postprocessing:
+        # Pre-processing always returns float32, because its output is consumed by the model.
+        # Post-processing, however, should return the dtype that is specified in the model spec.
+        # todo: cast dtype for inputs before preprocessing? or check dtype?
+        for out in outputs:
+            self._post.append(EnsureDtype(tensor_name=out.name, dtype=out.data_type))
+
+        self._req_input_stats = {s: self._collect_required_stats(self._prep, s) for s in SCOPES}
+        self._req_output_stats = {s: self._collect_required_stats(self._post, s) for s in SCOPES}
+        if any(self._req_output_stats[s] for s in SCOPES):
+            raise NotImplementedError("computing statistics for output tensors not yet implemented")
+
+        self._computed_dataset_stats: Optional[Dict[str, Dict[Measure, Any]]] = None
+
+        self.input_tensor_names = [ipt.name for ipt in inputs]
+        self.output_tensor_names = [out.name for out in outputs]
+        assert not any(name in self.output_tensor_names for name in self.input_tensor_names)
+        assert not any(name in self.input_tensor_names for name in self.output_tensor_names)
+
+    @property
+    def required_input_dataset_statistics(self) -> Dict[str, Set[Measure]]:
+        return self._req_input_stats[DATASET]
+
+    @property
+    def required_output_dataset_statistics(self) -> Dict[str, Set[Measure]]:
+        return self._req_output_stats[DATASET]
+
+    @property
+    def computed_dataset_statistics(self) -> Dict[str, Dict[Measure, Any]]:
+        return self._computed_dataset_stats
+
+    def apply_preprocessing(
+        self, *input_tensors: xr.DataArray
+    ) -> Tuple[List[xr.DataArray], Dict[str, Dict[Measure, Any]]]:
+        assert len(input_tensors) == len(self.input_tensor_names)
+        tensors = dict(zip(self.input_tensor_names, input_tensors))
+        sample_stats = self.compute_sample_statistics(tensors, self._req_input_stats[SAMPLE])
+        for proc in self._prep:
+            proc.set_computed_sample_statistics(sample_stats)
+            tensors[proc.tensor_name] = proc.apply(tensors[proc.tensor_name])
+
+        return [tensors[tn] for tn in self.input_tensor_names], sample_stats
+
+    def apply_postprocessing(
+        self, *output_tensors: xr.DataArray, input_sample_statistics: Dict[str, Dict[Measure, Any]]
+    ) -> Tuple[List[xr.DataArray], Dict[str, Dict[Measure, Any]]]:
+        assert len(output_tensors) == len(self.output_tensor_names)
+        tensors = dict(zip(self.output_tensor_names, output_tensors))
+        sample_stats = input_sample_statistics
+        sample_stats.update(self.compute_sample_statistics(tensors, self._req_output_stats[SAMPLE]))
+        for proc in self._post:
+            proc.set_computed_sample_statistics(sample_stats)
+            tensors[proc.tensor_name] = proc.apply(tensors[proc.tensor_name])
```

Comment:
I don't see where we make sure that the postprocessed tensors have the correct dtype.
(After postprocessing we need to return the same dtype as is specified in the output tensor spec.)

Reply:
here: https://github.com/bioimage-io/core-bioimage-io-python/pull/108/files#diff-1d7498945a046a703ef8456b221372224ab34bc5251067c2ba0bc7bf13025e5fR64-R69